### PR TITLE
Add/migration ssh credentials screen

### PR DESCRIPTION
**Pull request by jeroenpf**
Related to #77017 

## Proposed Changes

* Add a screen where SSH/FTP credentials can be entered before starting a migration.
* Do not ask for credentials if we found that the source site already has credentials configured.
* Show an `Upgrade your plan` page if the target site is not suitable for a migration.
* Show a helper menu along with the credentials forms to aid the user in finding the appropriate credentials.

## Testing Instructions

* Navigate to [http://calypso.localhost:3000/setup/import-focused/importerWordpress?siteSlug=${target_site}&from=${source_site}&option=everything](http://calypso.localhost:3000/setup/import-focused/importerWordpress?siteSlug=$%7Btarget_site%7D&from=$%7Bsource_site%7D&option=everything)
* Check with an Atomic site and verify that you see the `You are ready to migrate` page.
Screenshot 2566-05-17 at 20 03 31

* Check with a simple site and verify that you see the `Upgrade your plan` page.
Screenshot 2566-05-17 at 20 06 40

* Use a source site without a backup plan and verify that you can enter the credentials of the source site and proceed to migrate.
Screenshot 2566-05-17 at 20 04 36

* Use a source site with a backup plan or for which you previously provided credentials and ensure that the `You are ready to migrate` page does NOT display the credentials option.
Screenshot 2566-05-17 at 20 07 59

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Review by ouikhuan — APPROVED** [review]
Tested and work as described! Would love @bogiii to double check in case I miss anything else!

**Comment by javierarce** [comment]
While testing the initial screen of the flow (Zrk1NoQocqdk3uTphuuey4-fi-410_13411), I noticed that the fast migration option might be easy to overlook because the natural instinct is to click the button without reading the message. I hadn't anticipated this during the design phase, so I plan to improve that design.

For the time being, here are two suggestions to enhance the current page:

- I suggest we remove "entire" from `Your entire site {origin} will be migrated to {destination}` as it seems redundant.
- Rather than saying, `Optionally, provide the server credentials of your site to speed up the migration,` let's rephrase it to: `Want to speed up the migration? Provide the server credentials of your site`.

Regarding [the form](Zrk1NoQocqdk3uTphuuey4-fi-236_9951):

- The style of the labels (color) and input fields (border-radius, color, etc.) differs from other onboardings (check the newsletter one, for example).
- Same with the style of the link in the right column. Let's underline it and change its color to black (see: Zrk1NoQocqdk3uTphuuey4-fi-304_7545)
- I think we could auto-populate the `Server address` field and disable it
- The `Port` field doesn't need to be so wide.
- In other onboarding flows, we use a different pattern for required steps. For instance, in the newsletter setup, the submit button remains active, but we display an error message if the user fails to provide a newsletter name. We also follow this pattern in the importing stage of other onboardings. To maintain consistency, let's enable the continue button and show an error message if the user attempts to proceed to enter the credentials.
- The font size of the failed credentials error message ("We couldn't connect to your site. Please verify…"), should be `14px`.
- If the user inputs "root" as their username, we currently display this message: `We can't accept credentials for the root user. Please provide or create credentials for another user with access to your server.` Could we abbreviate this to just `We can't accept credentials for the root user.`?
- The `Only non-encrypted private keys are supported.` message under the text area should be aligned to the left.
- The error messages under the fields should be left-aligned

Screenshot 2023-05-18 at 11 14 44

File: client/blocks/importer/wordpress/import-everything/pre-migration/index.tsx [hunk 1]
```diff
@@ -0,0 +1,175 @@
+//create a pre-migration component
+
```
Comment (bogiii) [inline client/blocks/importer/wordpress/import-everything/pre-migration/index.tsx]: I just wanted to let you know that this comment and an empty line are unnecessary.

File: client/blocks/importer/wordpress/import-everything/pre-migration/index.tsx [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+import { Button } from '@automattic/components';
+import { SiteDetails } from '@automattic/data-stores';
+import { NextButton, SubTitle, Title } from '@automattic/onboarding';
+import classnames from 'classnames';
+import { useTranslate } from 'i18n-calypso';
+import React, { useEffect, useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { convertToFriendlyWebsiteName } from 'calypso/blocks/import/util';
+import MigrationCredentialsForm from 'calypso/blocks/importer/wordpress/import-everything/pre-migration/migration-credentials-form';
+import './style.scss';
```
Comment (bogiii) [inline client/blocks/importer/wordpress/import-everything/pre-migration/index.tsx]: We usually place the style import at the end of the `import` block.

File: client/blocks/importer/wordpress/import-everything/pre-migration/style.scss [hunk 3]
```diff
@@ -0,0 +1,118 @@
+@import "~calypso/blocks/import/style/base";
+@import "../style.scss";
+
+.import__pre-migration.import__import-everything--redesign {
+	max-width: 100%;
+}
+
+.import__pre-migration {
+
```
Comment (bogiii) [inline client/blocks/importer/wordpress/import-everything/pre-migration/style.scss]: No need for this empty line.

File: client/blocks/importer/wordpress/import-everything/pre-migration/style.scss [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	padding: 0 20px;
```
Comment (bogiii) [inline client/blocks/importer/wordpress/import-everything/pre-migration/style.scss]: Did you try using `rem` scss mixin for transforming px to rem?

This line could be like: `padding: 0 rem(20px)`.

For padding, margin, font-size, line-height, etc., we should rely on the rem unit.

File: client/blocks/importer/wordpress/import-everything/pre-migration/upgrade-plan.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+	return (
+		<>
+			<div
+				className={ classnames( 'import__import-everything', {
+					'import__import-everything--redesign': isEnabled( 'onboarding/import-redesign' ),
+				} ) }
+			>
+				<div className="import__heading-title">
+					<Title>Upgrade your plan</Title>
```
Comment (bogiii) [inline client/blocks/importer/wordpress/import-everything/pre-migration/upgrade-plan.tsx]: Missing translation part.

File: client/components/advanced-credentials/credentials-form/index.tsx [hunk 6]
```diff
@@ -375,13 +377,13 @@ const ServerCredentialsForm: FunctionComponent< Props > = ( {
 						) }
 				</FormFieldset>
 			) }
-			{ ! isAlternate && (
+			{ ! isAlternate && withHeader && (
 				<>
```
Comment (bogiii) [inline client/components/advanced-credentials/credentials-form/index.tsx]: No need for the `<>` fragment wrapper.

**Review by bogiii — APPROVED** [review]
I've left some comments for the minor changes.
I tested the regular import onboarding flow to check for some potential regression.

So far, so good!

**Comment by jeroenpf** [comment]
@javierarce 

> I suggest we remove "entire" from Your entire site {origin} will be migrated to {destination} as it seems redundant.
> Rather than saying, Optionally, provide the server credentials of your site to speed up the migration, let's rephrase it to: Want to speed up the migration? Provide the server credentials of your site.

I removed the from -> to line and rephrased the call to action for the credentials form.

> The style of the labels (color) and input fields (border-radius, color, etc.) differs from other onboardings (check the newsletter one, for example).

I applied the same style as is used for the form of the newsletter subscribers.

> I think we could auto-populate the Server address field and disable it

This is possible but the server address doesn't technically have to be the same as the domain name of the site. How about we prefill the hostname but leave it open for editing?

> The Port field doesn't need to be so wide.

The form is not designed by us. We are actually reusing the fairly complex credentials form and underlying logic for the jetpack backup credentials configuration. If I narrow the field, the label will be too long (especially when Bluehost is selected and an icon appears next to the label text). 

> In other onboarding flows, we use a different pattern for required steps. For instance, in the newsletter setup, the submit button remains active, but we display an error message if the user fails to provide a newsletter name. We also follow this pattern in the importing stage of other onboardings. To maintain consistency, let's enable the continue button and show an error message if the user attempts to proceed to enter the credentials.

I changed the button to remain active while the form is in 'error' state. If the user clicks the button while not all fields are filled in, an error will be shown.

> The font size of the failed credentials error message ("We couldn't connect to your site. Please verify…"), should be 14px.

Fixed.

> If the user inputs "root" as their username, we currently display this message: We can't accept credentials for the root user. Please provide or create credentials for another user with access to your server. Could we abbreviate this to just We can't accept credentials for the root user.?

Changed the original message by overriding the default form error.

> The Only non-encrypted private keys are supported. message under the text area should be aligned to the left.
> The error messages under the fields should be left-aligned

Both fixed.

**Comment by javierarce** [comment]
@jeroenpf:

Thanks for changes! 

> I applied the same style as is used for the form of the newsletter subscribers.

The margin-bottom is too tall, it should be `8px`. The font-size of the labels should be `16px`.

> This is possible but the server address doesn't technically have to be the same as the domain name of the site. How about we prefill the hostname but leave it open for editing?

Oh, ok! Let's do that then.

> The form is not designed by us. We are actually reusing the fairly complex credentials form and underlying logic for the jetpack backup credentials configuration. If I narrow the field, the label will be too long (especially when Bluehost is selected and an icon appears next to the label text).

Ok!

---

Checking everything again, I saw a few extra things: 

- We should center this error message at the bottom of the form:

Screenshot 2023-05-19 at 09 33 09

- The empty state of the fields is difficult to distinguish from the filled state. Could we match the color of the first field of [this form](https://wordpress.com/setup/newsletter/newsletterSetup). I think it's `color: var(--studio-gray-30);`


- To avoid having a multiple-line error message, let's say "Invalid port" instead of "Please enter a valid port number."

Screenshot 2023-05-19 at 09 42 38


- And for the error message under "WordPress installation path" let's say: 
"Please, enter a valid file path".
- If I remove the default port number and then I change to a different type of credential, the form populates the field with `NaN` and I can submit the form.
- If I use tabs to move through the fields, the focus goes to the help dropdown on the right column instead of the first input of the form. Could we fix that?
- When submitting invalid credentials I got this message: "Error saving. Please check your credentials and try again." Could we change the first sentence? It's not very clear what it means.

**Comment by jeroenpf** [comment]
> @jeroenpf:
> 
> 
> 
> Thanks for changes! 
> 
> 
> 
> > I applied the same style as is used for the form of the newsletter subscribers.
> 
> 
> 
> The margin-bottom is too tall, it should be `8px`. The font-size of the labels should be `16px`.
> 
> 
> 
> > This is possible but the server address doesn't technically have to be the same as the domain name of the site. How about we prefill the hostname but leave it open for editing?
> 
> 
> 
> Oh, ok! Let's do that then.
> 
> 
> 
> > The form is not designed by us. We are actually reusing the fairly complex credentials form and underlying logic for the jetpack backup credentials configuration. If I narrow the field, the label will be too long (especially when Bluehost is selected and an icon appears next to the label text).
> 
> 
> 
> Ok!
> 
> 
> 
> ---
> 
> 
> 
> Checking everything again, I saw a few extra things: 
> 
> 
> 
> - We should center this error message at the bottom of the form:
> 
> 
> 
> Screenshot 2023-05-19 at 09 33 09
> 
> 
> 
> - The empty state of the fields is difficult to distinguish from the filled state. Could we match the color of the first field of [this form](https://wordpress.com/setup/newsletter/newsletterSetup). I think it's `color: var(--studio-gray-30);`
> 
> 
> 
> 
> 
> - To avoid having a multiple-line error message, let's say "Invalid port" instead of "Please enter a valid port number."
> 
> 
> 
> Screenshot 2023-05-19 at 09 42 38
> 
> 
> 
> 
> 
> - And for the error message under "WordPress installation path" let's say: 
> 
> "Please, enter a valid file path".
> 
> - If I remove the default port number and then I change to a different type of credential, the form populates the field with `NaN` and I can submit the form.
> 
> - If I use tabs to move through the fields, the focus goes to the help dropdown on the right column instead of the first input of the form. Could we fix that?
> 
> - When submitting invalid credentials I got this message: "Error saving. Please check your credentials and try again." Could we change the first sentence? It's not very clear what it means. 
> 
> 

Thanks for all the feedback! All the proposed changes seem fine with me, I'll work on it.

File: client/blocks/importer/wordpress/import-everything/pre-migration/migration-credentials-form.tsx [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+		const credentials = { ...formState };
+
+		if ( formMode === FormMode.Password ) {
+			credentials.kpri = '';
+		} else if ( formMode === FormMode.PrivateKey ) {
+			credentials.pass = '';
+		}
+
+		// @todo custom event for migration flow?
+		dispatch( recordTracksEvent( 'calypso_jetpack_advanced_credentials_flow_credentials_update' ) );
```
Comment (bogiii) [inline client/blocks/importer/wordpress/import-everything/pre-migration/migration-credentials-form.tsx]: Does this event name suppose to be updated?

Something: `calypso_site_migration_credentials_update`

Comment (jeroenpf) [reply client/blocks/importer/wordpress/import-everything/pre-migration/migration-credentials-form.tsx]: Done! :-)

**Comment by jeroenpf** [comment]
@javierarce 

I fixed all of the issues you mentioned in the last feedback you gave. 

Additionally, I noticed some issues with the RTL display of the form for different breakpoints. Mainly these are regarding margins being added on the wrong side. I also fixed this.

There are two I can't (easily) solve:

> And for the error message under "WordPress installation path" let's say:
"Please, enter a valid file path".

This error is coming from a path validation library and I do not have control over the exact error message unless we implement our own path validator with custom errors which might not be worth it. These path errors are used throughout Calypso and can also be different if you for example try to add two slashes (/) or use a backslash (\).

> When submitting invalid credentials I got this message: "Error saving. Please check your credentials and try again." Could we change the first sentence? It's not very clear what it means.

This is also difficult to change as it is part of the original form and its underlying state management. This message seems to be triggered if the reason for saving the credentials fails is unclear.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8007910

Thank you @jeroenpf for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8007910).
